src/components/Courses.js

```diff
@@ -1,149 +1,128 @@
-    display: 'flex',
-    justifyContent: 'center'
+  actions: {
+    flexWrap: 'wrap',
+    justifyContent: 'center',
   },
-
-  empty: {
-    textAlign: 'center'
-  },
-
-  progress: {
-    margin: `0 ${theme.spacing(2)}px`
-  }
 }))
 
 export default function Courses(props) {
   const classes = useStyles()
-  const current_term = useSelector(state => state.terms.current_term)
-  const courses = useSelector(state => state.courses.courses)
-  const { mobile } = props
-  const dispatch = useDispatch();
-
-  useEffect(() => {
-    dispatch(fetch_courses(current_term))
-  }, [dispatch, current_term])
+  const courses = useSelector((state) => state.courses)
+  const courses_fetched = useSelector((state) => state.fetched)
+  const courses_fetching = useSelector((state) => state.fetching)
+  const courses_error = useSelector((state) => state.error)
 
-  if (courses === null || courses === []) {
+  if (courses_fetching === true) {
     return (
-      <Typography variant="h3" className={classes.empty} tabIndex="0">
+      <div className={classes.loading}>
+        <CircularProgress color='secondary' size={50} />
+      </div>
+    )
+  } else if (courses_fetched === true && courses_error === true) {
+    return (
+      <div className={classes.error}>
+        <ErrorMessages />
+      </div>
+    )
+  } else if (courses === null || courses === []) {
+    return (
+      <Typography variant='h3' className={classes.empty} tabIndex='0'>
         You currently have no courses for this semester.
       </Typography>
     )
-  }
-
-
-  let elements = []
-  for (let [i, course] of courses.entries()) {
-    if (course.meetings.length > 1 || course.instructors.length > 1) {
-      elements.push(
-        <ExpandableCourse
-          course={course}
-          key={'expandable' + Math.random()}
-          mobile={mobile}
-        />
-      )
-    } else if (!Object.is(course.waitList, '0')) {
-      elements.push(
-        <WaitlistCourse
-          course={course}
-          key={'waitlist' + Math.random()}
-          mobile={mobile}
-        />
-      )
-    } else {
-      elements.push(
-        <div
-          className={classes.courseContainer}
-          key={course.crn}
-        >
-          <div style={{ marginTop: '1em' }}>
-            <Card
-              className={classes.card}
-              key={course.crn + i + Math.random()}
-            >
-              <CourseHeader mobile={mobile} course={course} />
-              <CardContent
-                className={classes.content}
+  } else {
+    let elements = []
+    for (let [i, course] of courses.entries()) {
+      if (course.meetings.length > 1 || course.instructors.length > 1) {
+        elements.push(
+          <ExpandableCourse
+            course={course}
+            key={'expandable' + Math.random()}
+          />
+        )
+      } else if (!Object.is(course.waitlist, '0')) {
+        elements.push(
+          <WaitlistCourse course={course} key={'waitlist' + Math.random()} />
```

Comment:
Also, these `Math.random()` boyos. Maybe we tried this before and it wasn't unique enough, but could we do something like ```key={`waitlist-${course.crn}`}```?  It does work as-is, but it feels a bit hacky. And I know it was like that before, so I don't think it'll hold up this PR. But maybe as a follow-up issue

Reply:
i think i've also gotten rid of this in #212, a lot of components were combined so there wasn't the need for a ton of super unique keys. i'm using maps for everything so i'm just using the index as the key, and there hasn't seemed to be any console errors regarding duplicate/missing keys

Reply:
Cool, that sounds like a good solution to me :+1: